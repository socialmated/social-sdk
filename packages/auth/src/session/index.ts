export type * from './session.js';
export * from './oauth.js';
export * from './cookie.js';
