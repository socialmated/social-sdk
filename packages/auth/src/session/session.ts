import { type Promisable } from 'type-fest';

/**
 * Represents a session interface for managing session data, including retrieval,
 * refreshing, revocation, and expiration handling.
 */
export interface ReadonlySession {
  /**
   * Retrieves the value associated with the specified key from the session.
   *
   * @param key - The key whose value should be retrieved.
   * @returns A promise that resolves to the value as a string, or `null` if the key does not exist.
   */
  get: (key: string) => string | null;

  /**
   * Retrieves all key-value pairs stored in the session.
   *
   * @returns A promise that resolves to an object containing all session data.
   */
  getAll: () => Record<string, string>;

  /**
   * Refreshes the session, typically by updating access tokens or other session data.
   *
   * @param key - The key to refresh.
   *
   * @returns A promise that resolves to the refreshed value as a string.
   */
  refresh: (key?: string) => Promisable<string>;

  /**
   * Revokes the session, typically by invalidating access tokens or clearing session data.
   *
   * @returns A promise that resolves when the session is revoked.
   */
  revoke: () => Promisable<void>;

  /**
   * Get the remaining time in seconds until the session expires.
   *
   * @returns The remaining time in seconds until the session expires.
   */
  expiresIn: () => number;

  /**
   * Retrieves the scope of the session.
   *
   * @returns The scope as a string, or `undefined` if not found.
   */
  scope: () => string | undefined;

  /**
   * The issuer of the session, typically the base URL or domain.
   */
  issuer: URL | string;
}

export interface WritableSession extends ReadonlySession {
  /**
   * Sets a value in the session for the specified key.
   *
   * @param key - The key under which the value should be stored.
   * @param value - The value to store in the session.
   */
  set: (key: string, value: string) => void;

  /**
   * Clears the session data, effectively logging out the user or invalidating the session.
   */
  clear: () => void;
}
