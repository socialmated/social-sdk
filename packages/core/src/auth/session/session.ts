/**
 * Represents a session interface for managing session data, including retrieval,
 * refreshing, revocation, and expiration handling.
 */
export interface Session {
  /**
   * Retrieves the value associated with the specified key from the session.
   *
   * @param key - The key whose value should be retrieved.
   * @returns A promise that resolves to the value as a string, or `undefined` if not found.
   */
  get: (key: string) => string | undefined;

  /**
   * Retrieves all key-value pairs stored in the session.
   *
   * @returns A promise that resolves to an object containing all session data.
   */
  getAll: () => Record<string, string>;

  /**
   * Refreshes the session, typically by updating access tokens or other session data.
   *
   * @param key - The key to refresh. If not provided, refreshes all keys.
   *
   * @returns A promise that resolves when the session is refreshed.
   */
  refresh: (key?: string) => Promise<void>;

  /**
   * Revokes the session, typically by invalidating access tokens or clearing session data.
   *
   * @param key - The key to revoke. If not provided, revokes all keys.
   *
   * @returns A promise that resolves when the session is revoked.
   */
  revoke: (key?: string) => Promise<void>;

  /**
   * Get the remaining time in seconds until the session expires.
   *
   * @param key - The key to check for expiration.
   * @returns The remaining time in seconds until the session expires.
   */
  expiresIn: (key?: string) => number;

  /**
   * Retrieves the scope associated with the specified key from the session.
   *
   * @param key - The key whose scope should be retrieved. If not provided, retrieves the default scope.
   * @returns The scope as a string, or `undefined` if not found.
   */
  scope: (key?: string) => string | undefined;

  /**
   * Clears the session data, effectively logging out the user or invalidating the session.
   */
  clear: () => void;

  /**
   * The issuer of the session, typically the base URL or domain.
   */
  issuer: URL | string;
}
