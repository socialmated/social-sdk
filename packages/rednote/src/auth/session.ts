import {
  type CookieJar,
  type LocalStorage,
  type SessionStorage,
  WebStoreCookieSession,
} from '@social-sdk/auth/session';
import { addYear } from '@formkit/tempo';
import { generateLocalId, generateWebId } from '@/security/token/token.js';

export class RednoteCookieSession extends WebStoreCookieSession {
  /**
   * Creates an instance of the session class.
   *
   * @param url - The base URL for the session. Defaults to 'https://xiaohongshu.com' if not provided.
   * @param jar - An optional `CookieJar` instance to manage cookies for the session.
   * @param localStorage - An optional `LocalStorage` instance for managing local storage.
   * @param sessionStorage - An optional `SessionStorage` instance for managing session storage.
   */
  constructor(jar?: CookieJar, localStorage?: LocalStorage, sessionStorage?: SessionStorage) {
    super(new URL('https://www.xiaohongshu.com'), jar, localStorage, sessionStorage);
  }

  /**
   * Wraps a given `WebStoreCookieSession` into a `RednoteCookieSession` instance.
   *
   * @param session - The session object to wrap. Must have an issuer with hostname 'xiaohongshu.com'.
   * @returns A new instance of `RednoteCookieSession` initialized with the provided session's cookie jar, local storage, and session storage.
   * @throws If the session's issuer hostname is not 'xiaohongshu.com'.
   */
  public static wrap(session: WebStoreCookieSession): RednoteCookieSession {
    if (session.issuer.hostname !== 'www.xiaohongshu.com') {
      throw new Error('Invalid session host. Expected www.xiaohongshu.com.');
    }
    return new RednoteCookieSession(session.cookieJar, session.localStorage, session.sessionStorage);
  }

  /**
   * Refreshes and returns a specific session key value.
   *
   * @param key - The session key to refresh ('a1', 'webId', 'loadts', or 'xsecappid')
   * @returns Promise resolving to the refreshed key value
   */
  public override refresh(key: 'a1' | 'webId' | 'loadts' | 'xsecappid'): string {
    if (key === 'a1') {
      const localId = generateLocalId('Mac OS');
      this.set('a1', localId, {
        expires: addYear(new Date(), 1),
        path: '/',
        domain: '.xiaohongshu.com',
      });
      return localId;
    }
    if (key === 'loadts') {
      const ts = String(Date.now());
      this.set('loadts', ts, {
        expires: addYear(new Date(), 1),
        path: '/',
        domain: '.xiaohongshu.com',
      });
      return ts;
    }
    if (key === 'webId') {
      const webId = generateWebId(this.get('a1') ?? generateLocalId('Mac OS'));
      this.set('webId', webId, {
        expires: addYear(new Date(), 1),
        path: '/',
        domain: '.xiaohongshu.com',
      });
      return webId;
    }

    const appId = 'xhs-pc-web';
    this.set('xsecappid', appId, {
      expires: addYear(new Date(), 1),
      path: '/',
      domain: '.xiaohongshu.com',
    });
    return appId;
  }
}
