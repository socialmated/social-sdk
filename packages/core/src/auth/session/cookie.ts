import { canonicalDomain, Cookie, CookieJar } from 'tough-cookie';
import { type Page } from 'playwright';
import { LocalStorage, SessionStorage } from '@denostack/shim-webstore';
import { type Session } from './session.js';

/**
 * Manages session data using browser-like cookies.
 *
 * The `CookieSession` class provides methods to get, set, remove, and clear cookies
 * for a specific URL using a `CookieJar`. It implements the `Session` interface.
 *
 * @example
 * ```typescript
 * const session = new CookieSession(new URL('https://example.com'));
 * await session.set('token', 'abc123');
 * const token = await session.get('token');
 * await session.remove('token');
 * await session.clear();
 * ```
 *
 * @remarks
 * - Cookies are scoped to the provided URL's domain.
 * - Uses a `CookieJar` to manage cookies in-memory or via a persistent store.
 *
 * @see {@link Session}
 */
export class CookieSession implements Session {
  /**
   * Creates an instance of `CookieSession`.
   *
   * @param url - The URL for which cookies are managed.
   * @param cookieJar - An optional `CookieJar` instance for cookie management.
   */
  constructor(
    private url: URL | string,
    public cookieJar = new CookieJar(),
  ) {}

  /**
   * Creates a new `CookieSession` instance from the cookies present in the given Playwright `Page`.
   *
   * This method extracts all cookies from the page's browser context, converts them into `tough-cookie` `Cookie` objects,
   * and stores them in a `CookieJar`. The resulting `CookieSession` is initialized with the current page URL and the populated cookie jar.
   *
   * @param page - The Playwright `Page` instance from which to extract cookies.
   * @returns A promise that resolves to a new `CookieSession` containing the cookies from the page.
   */
  public static async fromPage(page: Page): Promise<CookieSession> {
    const url = page.url();

    // Get all cookies from the context
    const context = page.context();
    const cookies = await context.cookies(url);

    // Create a new CookieJar to store the cookies
    const cookieJar = new CookieJar();

    // Iterate over the cookies and add them to the jar
    for (const cookie of cookies) {
      const toughCookie = new Cookie({
        key: cookie.name,
        value: cookie.value,
        domain: canonicalDomain(cookie.domain), // Normalize domain
        path: cookie.path,
        expires: cookie.expires ? new Date(cookie.expires * 1000) : undefined,
        httpOnly: cookie.httpOnly,
        secure: cookie.secure,
        sameSite: cookie.sameSite,
      });

      // Set the cookie in the jar, ignoring errors for invalid cookies
      await cookieJar.setCookie(toughCookie, url.toString(), {
        ignoreError: true,
      });
    }

    return new CookieSession(url, cookieJar);
  }

  /**
   * Retrieves the value of a cookie by its key from the current cookie jar.
   *
   * @param key - The name of the cookie to retrieve.
   * @returns The value of the cookie as a string, or `undefined` if the cookie does not exist.
   */
  public get(key: string): string | undefined {
    const found = this.cookieJar.getCookiesSync(this.issuer.toString(), {
      expire: false,
      allPaths: true,
    });
    const cookie = found.find((c) => c.key === key);

    return cookie?.value;
  }

  /**
   * Retrieves all cookies associated with the current URL as a key-value map.
   *
   * @returns An object containing all cookies, where the keys are cookie names and the values are cookie values.
   */
  public getAll(): Record<string, string> {
    const cookies = this.cookieJar.getCookiesSync(this.issuer.toString(), {
      expire: false,
      allPaths: true,
    });
    const cookieMap: Record<string, string> = {};
    cookies.forEach((cookie) => {
      cookieMap[cookie.key] = cookie.value;
    });

    return cookieMap;
  }

  /**
   * Removes all cookies from the internal cookie jar, effectively clearing any stored session data.
   */
  public clear(): void {
    this.cookieJar.removeAllCookiesSync();
  }

  /**
   * Refreshes the session by updating the cookies in the cookie jar.
   *
   * @remarks
   * This method is a placeholder for any additional logic that may be needed to refresh the session.
   *
   * @returns A promise that resolves to the refreshed cookie value.
   */
  public refresh(_?: string): Promise<string> {
    return Promise.reject(new Error('Not implemented.'));
  }

  /**
   * Revokes the session by clearing all cookies from the cookie jar.
   *
   * @remarks
   * This method is a placeholder for any additional logic that may be needed to revoke the session.
   *
   * @returns A promise that resolves when the revoke operation is complete.
   */
  public async revoke(_?: string): Promise<void> {
    // Implement any necessary logic to revoke the session if needed
    return Promise.reject(new Error('Not implemented.'));
  }

  /**
   * Retrieves the remaining time in seconds until the specified cookie expires.
   *
   * @param key - The key of the cookie to check for expiration.
   * @returns The remaining time in seconds until the cookie expires, or `NaN` if the cookie does not exist.
   */
  public expiresIn(key?: string): number {
    const cookies = this.cookieJar.getCookiesSync(this.issuer.toString(), {
      expire: false,
      allPaths: true,
    });

    const found = cookies.find((c) => c.key === key);
    if (!found) {
      return -Infinity;
    }
    if (!found.expires) {
      return NaN;
    }
    if (found.expires === 'Infinity') {
      return Infinity;
    }
    return Math.floor((found.expires.getTime() - Date.now()) / 1000);
  }

  /**
   * Retrieves the path of a cookie with the specified key from the cookie jar.
   *
   * @param key - The key of the cookie to search for. If not provided, the method will return `undefined`.
   * @returns The path of the found cookie, or `undefined` if no cookie with the given key exists.
   */
  public scope(key?: string): string | undefined {
    const cookies = this.cookieJar.getCookiesSync(this.issuer.toString(), {
      expire: false,
      allPaths: true,
    });

    const found = cookies.find((c) => c.key === key);

    return found?.path ?? undefined;
  }

  /**
   * The issuer of the session, typically the base URL or domain.
   */
  public get issuer(): URL {
    return new URL(this.url);
  }
}

/**
 * Extends {@link CookieSession} to provide unified management of cookies, localStorage, and sessionStorage
 * for web sessions, with persistent storage for localStorage data.
 *
 * @remarks
 * - LocalStorage data is persisted to disk in a file path based on the issuer's hostname.
 * - SessionStorage data is managed in-memory for the duration of the session.
 *
 * @example
 * ```typescript
 * const session = await WebStoreCookieSession.fromPage(page);
 * const token = session.get('authToken', 'local');
 * session.clear();
 * ```
 *
 * @public
 */
export class WebStoreCookieSession extends CookieSession {
  /**
   * Creates an instance of `WebStoreCookieSession`.
   *
   * @param url - The URL for which cookies are managed.
   * @param cookieJar - An optional `CookieJar` instance for cookie management.
   * @param localStorage - An optional `LocalStorage` instance for managing local storage.
   * @param sessionStorage - An optional `SessionStorage` instance for managing session storage.
   */
  constructor(
    url: URL | string,
    cookieJar?: CookieJar,
    public localStorage = new LocalStorage(`.data/${new URL(url).hostname}/localstorage.json`),
    public sessionStorage = new SessionStorage(),
  ) {
    super(url, cookieJar);
  }

  /**
   * Creates a {@link WebStoreCookieSession} instance from the given Playwright {@link Page}.
   *
   * This method extracts the session information, localStorage, and sessionStorage data
   * from the provided page. It persists the localStorage data to a file based on the session's
   * issuer hostname and assigns the sessionStorage data directly.
   *
   * @param page - The Playwright {@link Page} from which to extract session and storage data.
   * @returns A promise that resolves to a {@link WebStoreCookieSession} containing the session,
   *          localStorage, and sessionStorage data.
   */
  public static override async fromPage(page: Page): Promise<WebStoreCookieSession> {
    const session = await CookieSession.fromPage(page);
    const localStorage = new LocalStorage(`.data/${new URL(session.issuer).hostname}/localstorage.json`);
    const sessionStorage = new SessionStorage();

    const storageState = await page.context().storageState();
    const localStorageData =
      storageState.origins
        .find((origin) => origin.origin === session.issuer.origin)
        ?.localStorage.reduce(
          (acc, item) => {
            acc[item.name] = item.value;
            return acc;
          },
          {} as Record<string, string>,
        ) ?? {};
    localStorage._persistData(localStorageData);

    const sessionStorageData = await page.evaluate(() => window.sessionStorage);
    sessionStorage._data = sessionStorageData;

    return new WebStoreCookieSession(session.issuer, session.cookieJar, localStorage, sessionStorage);
  }

  /**
   * Retrieves the value associated with the specified key from the chosen storage type.
   *
   * @param key - The key whose value should be retrieved.
   * @param from - The storage type to retrieve the value from. Can be `'cookie'`, `'local'`, or `'session'`. Defaults to `'cookie'`.
   * @returns The value associated with the key if it exists, otherwise `undefined`.
   */
  public override get(key: string, from: 'cookie' | 'local' | 'session' = 'cookie'): string | undefined {
    if (from === 'cookie') {
      return super.get(key);
    }
    if (from === 'local') {
      return this.localStorage.getItem(key) ?? undefined;
    }
    return this.sessionStorage.getItem(key) ?? undefined;
  }

  /**
   * Retrieves all stored session data from the specified storage type.
   *
   * @param from - The storage type to retrieve data from. Can be 'cookie', 'local', or 'session'.
   * @returns An object containing all key-value pairs from the selected storage.
   */
  public override getAll(from: 'cookie' | 'local' | 'session' = 'cookie'): Record<string, string> {
    if (from === 'cookie') {
      return super.getAll();
    }
    if (from === 'local') {
      return this.localStorage._readData();
    }
    return this.sessionStorage._data;
  }

  /**
   * Clears all data from both localStorage and sessionStorage, effectively removing all stored session data.
   */
  public override clear(): void {
    super.clear();
    this.localStorage.clear();
    this.sessionStorage.clear();
  }
}
