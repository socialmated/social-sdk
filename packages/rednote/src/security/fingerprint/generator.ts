import { type Options } from 'got';
import { gotScraping } from '@social-sdk/client/http';
import { xsCommon } from './fingerprint.js';
import { type RednoteCookieSession } from '@/auth/session.js';
import { type ApiResponse } from '@/types/common.js';
import { type SbtSource } from '@/types/security.js';

/**
 * XSCommonGenerator is responsible for generating a common fingerprint signature
 * used for security and anti-abuse mechanisms in the Xiaohongshu (XHS) platform.
 *
 * @example
 * ```typescript
 * const generator = new XSCommonGenerator(session);
 * const signature = generator.generate(config, requestOptions);
 * ```
 *
 * @public
 */
export class XSCommonGenerator {
  /**
   * Creates an instance of the fingerprint generator.
   *
   * @param session - The current Rednote cookie session used for fingerprint generation.
   * @param commonPatterns - An optional array of common API endpoint patterns that should trigger fingerprint generation.
   */
  constructor(
    private session: RednoteCookieSession,
    private commonPatterns: (string | RegExp)[],
  ) {}

  /**
   * Retrieves the current signature count from session storage.
   * Optionally increments the count before returning it.
   *
   * @param increment - If true, increments the signature count before returning.
   * @returns The current (or incremented) signature count as a number.
   */
  private getSigCount(increment?: boolean): number {
    const sessionStorage = this.session.sessionStorage;
    const key = 'sc';

    let count = Number(sessionStorage.getItem(key)) || 0;
    if (increment) {
      count++;
      sessionStorage.setItem(key, String(count));
    }
    return count;
  }

  /**
   * Updates the `commonPatterns` array with the provided patch values.
   *
   * @param patch - An array of string patterns to be added to `commonPatterns`.
   * @returns A Promise that resolves when the operation is complete.
   */
  private async pushXsCommon(): Promise<void> {
    const sbtSource = await gotScraping
      .post('https://as.xiaohongshu.com/api/sec/v1/sbtsource', {
        headers: {
          'sec-fetch-dest': 'empty',
          'sec-fetch-mode': 'cors',
          'sec-fetch-site': 'same-site',
        },
      })
      .json<ApiResponse<SbtSource>>();

    const commonPatch = sbtSource.data?.commonPatch.length
      ? sbtSource.data.commonPatch
      : ['api/sec/v1/shield/webprofile'];

    commonPatch.forEach((c) => {
      this.commonPatterns.push(c);
    });
  }

  /**
   * Generates a fingerprint string based on the provided client configuration and request options.
   *
   * @param config - The client configuration object containing platform information.
   * @param request - The request options, including URL and headers.
   * @returns The generated fingerprint string, or `null` if the URL does not match path patterns.
   */
  public generate(platform: string, request: Options): string | null {
    const localStorage = this.session.localStorage;

    const url = request.url ? request.url.toString() : '';
    if (!this.commonPatterns.map((p) => new RegExp(p)).some((pattern) => pattern.test(url))) {
      return null;
    }

    const xt = (request.headers['X-t'] ?? '') as string;
    const xs = (request.headers['X-s'] ?? '') as string;
    const xSign = (request.headers['X-sign'] ?? '') as string;

    const shouldIncrement = Boolean((xt && xs) || xSign);
    const sc = this.getSigCount(shouldIncrement);

    const a1 = this.session.get('a1') ?? '';
    const b1 = localStorage.getItem('b1') ?? '';
    const b1b1 = localStorage.getItem('b1b1') ?? '1';

    return xsCommon(platform, a1, b1, b1b1, sc);
  }
}
