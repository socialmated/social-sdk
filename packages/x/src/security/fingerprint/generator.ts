import 'golang-wasm-exec';
import { loadWasmModule } from './wasm.js';
import { type XCookieSession } from '@/auth/session.js';

declare global {
  /**
   * The global function that will be set by the WebAssembly module to generate the XP-Forwarded-For string.
   */
  var getForwardedForStr: (() => Promise<{ str: string; expiryTimeMillis: string }>) | undefined;
}

/**
 * Generates a XP-Forwarded-For string using a WebAssembly module.
 *
 * @example
 * ```typescript
 * const session = new XCookieSession();
 * const generator = new XPFwdForGenerator(session);
 * const result = await generator.generate();
 * console.log(result.str); // The generated XP-Forwarded-For string
 * console.log(result.expiryTimeMillis); // The expiry time in milliseconds
 * ```
 */
export class XPFwdForGenerator {
  /**
   * Creates an instance of XPFwdForGenerator.
   * This constructor initializes the global navigator and document properties
   * to simulate a browser environment for the WebAssembly module.
   *
   * @param session - The XCookieSession used to retrieve cookies for the request.
   */
  constructor(session: XCookieSession) {
    Object.assign(globalThis.navigator, {
      userActivation: {
        hasBeenActive: false,
      },
    });
    Object.defineProperty(globalThis.navigator, 'userAgent', {
      get: () =>
        'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/137.0.0.0 Safari/537.36',
      configurable: true,
    });
    Object.defineProperty(globalThis, 'document', {
      value: {
        cookie: session.cookieJar.getCookieStringSync(session.issuer.toString()),
      },
    });
  }

  /**
   * Generates a XP-Forwarded-For string using the WebAssembly module.
   *
   * @returns A promise that resolves to an object containing the generated XP-Forwarded-For string and its expiry time in milliseconds.
   */
  public async generate(): Promise<{
    str: string;
    expiryTimeMillis: string;
  }> {
    if (!globalThis.getForwardedForStr) {
      await this.init();
    }

    if (typeof globalThis.getForwardedForStr === 'function') {
      const res = await globalThis.getForwardedForStr();
      return res;
    }

    throw new Error('Wasm module did not set window.getForwardedForStr');
  }

  /**
   * Initializes the WebAssembly module and sets up the global getForwardedForStr function.
   * This method must be called before generating the XP-Forwarded-For string.
   *
   * @returns A promise that resolves when the initialization is complete.
   */
  private async init(): Promise<void> {
    const go = new Go();
    try {
      const module = await loadWasmModule({
        ...go.importObject,
        env: {
          ...go.importObject['env'],
          memory: go.importObject['env']?.['memory'] ?? new WebAssembly.Memory({ initial: 10 }),
          table:
            go.importObject['env']?.['table'] ??
            new WebAssembly.Table({
              initial: 0,
              element: 'anyfunc',
            }),
        },
      });

      void go.run(module.instance);
    } catch {
      /* empty */
    }
  }
}
