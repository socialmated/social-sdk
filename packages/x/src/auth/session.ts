import { CookieSession } from '@social-sdk/core/auth/session';
import { type CookieJar } from 'tough-cookie';
import { addHour, addYear } from '@formkit/tempo';
import { fetchGuestToken, generateCsrfToken } from '../security/token/index.js';

/**
 * Represents a session for the X platform (formerly Twitter) using cookies.
 */
export class XCookieSession extends CookieSession {
  /**
   * Creates an instance of the session class.
   *
   * @param url - The base URL for the session. Defaults to 'https://x.com' if not provided.
   * @param jar - An optional `CookieJar` instance to manage cookies for the session.
   */
  constructor(jar?: CookieJar) {
    super(new URL('https://x.com'), jar);
  }

  /**
   * Wraps a given `CookieSession` into an `XCookieSession` instance.
   *
   * @param session - The `CookieSession` object to be wrapped.
   * @returns An `XCookieSession` instance created from the provided `CookieSession`.
   * @throws If the session's issuer hostname is not `x.com`.
   */
  public static wrap(session: CookieSession): XCookieSession {
    if (session.issuer.hostname !== 'x.com') {
      throw new Error('Invalid session host. Expected x.com.');
    }
    return new XCookieSession(session.cookieJar);
  }

  /**
   * Refreshes the session by updating the guest token.
   *
   * @param key - The key to refresh. Defaults to 'gt' (guest token).
   * @returns A promise that resolves when the guest token has been successfully refreshed.
   */
  public override async refresh(key: 'gt' | 'ct0'): Promise<string> {
    if (key === 'ct0') {
      const csrfToken = generateCsrfToken();
      this.set('ct0', csrfToken, {
        expires: addYear(new Date(), 1),
        path: '/',
        domain: '.x.com',
        secure: true,
        sameSite: 'Lax',
      });
      return csrfToken;
    }

    const guestToken = await fetchGuestToken();
    this.set('gt', guestToken, {
      expires: addHour(new Date(), 2.5),
      path: '/',
      domain: '.x.com',
      secure: true,
      sameSite: 'None',
    });
    return guestToken;
  }
}
