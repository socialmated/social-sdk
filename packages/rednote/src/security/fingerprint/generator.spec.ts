import { Options } from 'got';
import { XSCommonGenerator } from './generator.js';
import { RednoteCookieSession } from '@/auth/session.js';

describe(XSCommonGenerator, () => {
  const a1 = '1972697fa3d2ly0o5nfy1iqg0ra0w56q79gz2gm0450000417121';
  const b1 =
    'I38rHdgsjopgIvesdVwgIC+oIELmBZ5e3VwXLgFTIxS3bqwErFeexd0ekncAzMFYnqthIhJeSBMDKutRI3KsYorWHPtGrbV0P9WfIi/eWc6eYqtyQApPI37ekmR6QL+5Ii6sdneeSfqYHqwl2qt5B0DBIx+PGDi/sVtkIxdsxuwr4qtiIhuaIE3e3LV0I3VTIC7e0utl2ADmsLveDSKsSPw5IEvsiVtJOqw8BuwfPpdeTFWOIx4TIiu6ZPwrPut5IvlaLbgs3qtxIxes1VwHIkumIkIyejgsY/WTge7eSqte/D7sDcpipedeYrDtIC6eDVw2IENsSqtlnlSuNjVtIx5e1qt3bmAeVn8LIESLIEk8+9DUIvzy4I8OIic7ZPwFIviR4o/sDLds6PwVIC7eSd7sf0k4IEve6WGMtVwUIids3s/sxZNeiVtbcUeeYVwRIvM/z06eSuwvgf7sSqweIxltIxZSouwOgVwpsoTHPW5ef7NekuwcIEosSgoe1LuMIiNeWL0sxdh5IiJsxPw9IhR9JPwJPutWIv3e1Vt1IiNs1qw5IEKsdVtFtuw4sqwFIvhvIxqzGniRKWoexVtUIhW4Ii0edqwpBlb2peJsWU4TIiGb4PtOsqwEIvNexutd+pdeVYdsVDEbIhos3odskqt8pqwQIvNeSPwvIieeT/ubIveeSBveDPtXIx0sVqw64B8qIkWJIvvsxFOekaKsDYeeSqwoIkpgIEpYzPwqIxGSIE7eirqSwnvs0VtZIhpBbut14lNedM0eYPwpmPwZIC+7IiGy/VwttVtaIC5e0pesVPwFJqwBIhW=';
  let session: RednoteCookieSession;
  let generator: XSCommonGenerator;

  beforeEach(() => {
    session = new RednoteCookieSession();
    session.set('a1', a1);
    session.set('b1', b1, { to: 'local' });
    session.set('b1b1', '1', { to: 'local' });
    session.set('sc', '10', { to: 'session' });

    generator = new XSCommonGenerator(session, ['/api/sns/web']);
  });

  describe(XSCommonGenerator.prototype.generate, () => {
    it('should generate X-S-Common', () => {
      const req = new Options('https://edith.xiaohongshu.com/api/sns/web/v1/resource');
      const result = generator.generate('Mac OS', req);

      expect(result).toMatchSnapshot();
    });
  });
});
