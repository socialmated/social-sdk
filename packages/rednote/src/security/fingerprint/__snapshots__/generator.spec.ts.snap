// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`XSCommonGenerator > generate > should generate X-S-Common 1`] = `"2UQAPsHCPUIjqArjwjHjNsQhPsHCH0rjNsQhPaHCH0c1PahIHjIj2eHjwjQ+GnPW/MPjNsQhPUHCHdYiqUMIGUM78nHjNsQh+sHCH0c1+0q1PsHVHdWMH0ijP/D7P0GE+98YP9cUJoDIJAp18dDlygbdPoQYPoqM+dr7wn4CPf4TPecMPeZIPecl+ArUPaHVHdW9H0ijHjIj2eqjwjHjNsQhwsHCHDDAwoQH8B4AyfRI8FS98g+Dpd4daLP3JFSb/BMsn0pSPM87nrldzSzQ2bPAGdb7zgQB8nph8emSy9E0cgk+zSS1qgzianYt8p+s/LzN4gzaa/+NqMS6qS4HLozoqfQnPbZEp98QyaRSp9P98pSl4oSzcgmca/P78nTTL08z/sVManD9q9z18np/8db8aob7JeQl4epsPrzsagW3Lr4ryaRApdz3agYDq7YM47HFqgzkanYMGLSbP9LA/bGIa/+nprSe+9LI4gzVPDbrJg+P4fprLFTALMm7+LSb4d+kpdzt/7b7wrQM498cqBzSpr8g/FSh+bzQygL9nSm7qSmM4epQ4flY/BQdqA+l4oYQ2BpAPp87arS34nMQyFSE8nkdqMD6pMzd8/4SL7bF8aRr+7+rG7mkqBpD8pSUzozQcA8Szb87PDSb/d+/qgzVJfl/4LExpdzQ2epSPgbFP9QTcnpnJ0YPaLp//rSbyAW3wLzpag8C2/zQwrRQynP7nSm7zDS9ypHFJURAzrlDqA8c4M8QcA4SL9c7q9GIyAzQzg8S+S4o/gzn4MpQynzAP7P6q7Yy/fpkpdzjGMpS8pSn4MQQ4DF620Z98p+M478d804AL7b78LShJozQ2bk/J7p7/94n47mAJMzHLbqM8nG7/fp34g40aLp6qM+dJ9Ll/op+anSw8p4PPo+h8BWManStq7Yc4ASQybHEaSm7aSmM4b4Q40+SPp8FPLSk/dPlqgqMaLpNq9zn4r8F4gqFq7b7zDS9yo8Q2obCz9EkLDTgJ9phpdzpanYg+rSkPBpDqg4IcfljPdmSad+gp/zLanSoG0zc4rRAqg4bag8w8gYM4Bc3qBzSpSSDqM8rznQQyBRAP9RDq9Tl4eYIqg4zag8w8p+c478QynpSpsRMGDS98np/cd8SzbmFnrShPo+nqgq9+rHhqLS3pFkQ4d8A2r8O8nTYa7+rnnpSL7b7JFS3qB4Qzgm82Sm7qLShzM+Qz/4SygQlL7414dPIpdzyanYIcfQM4erFJrES8rFI8pSc47mTLo4yaLP3+FSkz7D6pd4F4b8FGLSe+nLIqBpApSm7zDkl4FQQybqRHjIj2eDjwjFM+0Pl+0cUw/qVHdWlPsHCP/ZVHdWlPaHCHfE6qfMYJsQR"`;
